WIP: Adds decorators for data helpers

p-galligan opened the pull request:
## Proposed Changes

  - Add more decorators to cover more object types.

Commit:
add dictionary checking

Adds a decorator to check whether the argument passed to a function is a dictionary.

Commit:
separate decorators

Put decorator functions into a separate file

helrond commented on rac_aspace/decorators.py:
```diff
@@ -0,0 +1,9 @@
+
+def check_dictionary(func):
+    def inner(*args, **kwargs):
+        if not isinstance(args, dict):
```
```suggestion
        if not isinstance(args[0], dict):
```
`args` is a tuple containing all the positional arguments passed to the function, so we'll need to get the first one OR alternatively, check to make sure ALL of the args are dicts, which (I think) would look like:

```suggestion
        if not [isinstance(a, dict) for a in args]:
```

helrond commented on rac_aspace/decorators.py:
```diff
@@ -0,0 +1,9 @@
+    def inner(*args, **kwargs):
+        if not isinstance(args, dict):
+            print('Parameter is not a dictionary')
+            return
```
```suggestion
            raise TypeException('{} is not a dictionary'.format(args[0])
```
Let's `raise` rather than simply printing so we can stop bad things from happening.

helrond reviewed (commented):
I think this is a great demo of how decorators can work, and sets up future work on this nicely. I've added some suggested changes.

As we get further down the line, we may be able to factor checking for multiple types into a single decorator by passing the expected type. But we can figure that out as we get further along.

p-galligan committed:
Update rac_aspace/decorators.py

Co-Authored-By: Hillel Arnold <helrond@hotmail.com>

p-galligan committed:
Update rac_aspace/decorators.py

Co-Authored-By: Hillel Arnold <helrond@hotmail.com>

helrond commented on rac_aspace/decorators.py:
```diff
@@ -0,0 +1,8 @@
+def check_dictionary(func):
+    def inner(*args, **kwargs):
+        if not [isinstance(a, dict) for a in args]:
+            raise TypeException('{} is not a dictionary'.format(args[0])
```
```suggestion
            raise TypeException('{} is not a dictionary'.format(args[0]))
```
sorryyyyyyy

p-galligan replied:
I should have checked closer.

p-galligan committed:
Update rac_aspace/decorators.py

Co-Authored-By: Hillel Arnold <helrond@hotmail.com>

helrond commented:
Perhaps overly pedantic, but this PR doesn't actually fully address #37 so perhaps that `fixes` should be removed from the description?

helrond committed:
target correct exception

Commit:
adding to decorators

Adding some functionality to decorators.

Commit:
fix formatting

forgot to run tox first

Commit:
check and validate for strings

adding decorators to check and validate for strings

Commit:
adding for loop

adding for loops to validate decorators

Commit:
fixing logic

fixing validate logic

Commit:
adding decorators

adding decorators and removing unused helpers

Commit:
linting changes

getting linting to pass

Commit:
updating requirements

updating pre-commit requirements

p-galligan committed:
Merge branch 'master' into issue-37

p-galligan committed:
Merge branch 'master' into issue-37

helrond commented:
I have a couple thoughts on this that are more conversations than change requests:
- It looks like each of `check_dictionary`, `check_list` and `check_str` functions check *all* the arguments passed to a function. In a number of cases (`text_in_note()`,`format_resource_id()`) we're going to have arguments of different types passed in so those decorators are going to throw an exception. 
  - One solution to this might be to have the decorator target only the first argument, and then *always* pass the main data object as the first argument. Thoughts?
- I'm wondering if we can refactor `check_dictionary`, `check_list` and `check_str` into a single `@check_type` (or similar) decorator which takes the expected type as an argument. So you'd then use it like `@check_type(str)`.
- I'd like to hear a bit more about the idea behind the `validate_*` decorators. It seems like they're sort of doing what unit tests should be doing?

p-galligan commented:
For the `validate` decorators, I had some questions about where decorators end and where tests begin. So, I could check instances in tests, but that could also be done in decorators.

I wasn't really able to find if Python would rather you do it one way, but it might make more sense in tests.

helrond commented:
IMO, my understanding of the point of decorators is to provide meaningful error messages to users about things that might not be obvious to them (such as "hey, you need to pass an object into this function"), not to ensure that the output of functions is correct. From a user perspective, a decorator which validates the output doesn't really help them that much; they'll likely just check the output themselves if they don't know what to expect.

Commit:
remove validation decorators

Removing unnecessary decorators

Commit:
remove imports

forgot to remove the imports because I ate too big a lunch

Commit:
close the parenthesis

close that up

p-galligan commented:
@helrond I've been trying to think about ways to get these decorators to target just the first argument, or to find examples of how that might work, and I'm having some difficulty. Do you have any reading or examples I might be able to look at?

helrond commented:
`args` is a list, so to just get the first one, you'd do `args[0]`. Is there something more complex you want to do?

Commit:
update decorators for first arg

Updating decorators to only check first argument. Something like this @helrond ?

Commit:
fix linting

forgot to remove a bracket

Commit:
Revert "fix linting"

This reverts commit 561bfdc54b150e37363fc31d65cefb8fc21cb9b9.

Commit:
Revert "update decorators for first arg"

This reverts commit 64e3bfca1cb3e269645177f03d15c8c9cd440843.

Commit:
update decorators

update decorators and data_helpers

Commit:
remove check_str

remove check_str from data_helpers import because it is unused

Commit:
update data helpers

updating data helpers file to bring in line with master

p-galligan committed:
Merge branch 'master' into issue-37

Commit:
fix linting

fixing spacing

Commit:
remove decorators from functions

removing decorators from functions for now

Commit:
fix linting

adding a space to make a block comment

Commit:
add check_str

adding check_str back in because it still works

helrond committed:
refactors decorators, updates docstrings

p-galligan committed:
Merge branch 'master' into issue-37

Commit:
Update decorator wrappers

Updating some types in decorator wrappers.

Commit:
remove wrapper

removing wrapper from get_orphans

p-galligan commented:
@helrond I removed the wrapper from `get_orphans` because it was failing as a list, as well as some others that I tried (JSONModelObject and dict).

I'm happy to put it back, but I'm not sure what type it actually should be.

`_____________________________________________________________________________________ TestDataHelpers.test_get_orphans _____________________________________________________________________________________

self = <tests.test_data_helpers.TestDataHelpers testMethod=test_get_orphans>

    def test_get_orphans(self):
        with rac_vcr.use_cassette("test_get_orphans.json"):
            repository = ASpace(
                baseurl="http://localhost:8089").repositories(2)
            archival_objects = repository.archival_objects
            orphans = data_helpers.get_orphans(
>               archival_objects, "linked_agents")

tests/test_data_helpers.py:80: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (#<JSONModelRelation:/repositories/2/archival_objects:{}>, 'linked_agents'), kwargs = {}

    def wrapper(*args, **kwargs):
        if not isinstance(args[0], obj_type):
>           raise TypeError("{} is not a {}".format(args[0], obj_type))
E           TypeError: #<JSONModelRelation:/repositories/2/archival_objects:{}> is not a <class 'list'>

rac_aspace/decorators.py:5: TypeError`

helrond commented:
@p-galligan I think that's fine - it's probably a generator object (you could confirm that by doing `print(type(object_list))` but I think it's fine to leave it as-is.
